File: apps/dot/backend/app/routers/contacts.py
"""Router de agenda local (phonebook) — UI Configuración."""
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.auth_deps import claims_uid, require_product_jwt
from app.services import contacts_store as store

router = APIRouter(prefix="/v1/contacts", tags=["contacts"])


class ContactCreateBody(BaseModel):
    name: str = Field(min_length=1)
    phone: str = ""
    email: str = ""
    notes: str = ""


def _serialize_contact(contact: dict[str, Any]) -> dict[str, Any]:
    return {
        "id": contact.get("id"),
        "name": contact.get("name"),
        "phone": contact.get("phone"),
        "email": contact.get("email"),
        "tags": contact.get("tags") or [],
        "notes": contact.get("notes") or "",
        "source": contact.get("source") or "manual",
        "updated_at": contact.get("updated_at"),
    }


@router.get("")
def list_contacts(claims: dict = Depends(require_product_jwt)):
    _ = claims_uid(claims)
    contacts = store.read_contacts()
    return {
        "total": len(contacts),
        "contacts": [_serialize_contact(c) for c in contacts],
    }


@router.post("")
def create_contact(body: ContactCreateBody, claims: dict = Depends(require_product_jwt)):
    _ = claims_uid(claims)
    ok, contact, message = store.add_contact(
        name=body.name,
        phone=body.phone,
        email=body.email,
        notes=body.notes,
        source="manual",
    )
    if not ok:
        raise HTTPException(status_code=503, detail=message)
    return {"ok": True, "message": message, "contact": _serialize_contact(contact or {})}


@router.delete("/{contact_id}")
def remove_contact(contact_id: str, claims: dict = Depends(require_product_jwt)):
    _ = claims_uid(claims)
    ok, message = store.delete_contact(contact_id)
    if not ok:
        raise HTTPException(status_code=404 if "no encontrado" in message.lower() else 503, detail=message)
    return {"ok": True, "message": message}


@router.post("/import/gmail")
def import_gmail_contacts(claims: dict = Depends(require_product_jwt)):
    uid = claims_uid(claims)
    ok, message = store.import_from_gmail(uid)
    if not ok:
        raise HTTPException(status_code=503, detail=message)
    contacts = store.read_contacts()
    return {"ok": True, "message": message, "total": len(contacts)}


@router.post("/import/whatsapp")
def import_whatsapp_contacts(claims: dict = Depends(require_product_jwt)):
    uid = claims_uid(claims)
    ok, message = store.import_from_whatsapp(uid)
    if not ok:
        raise HTTPException(status_code=503, detail=message)
    contacts = store.read_contacts()
    return {"ok": True, "message": message, "total": len(contacts)}

File: apps/dot/frontend/src/lib/api/contacts.ts
import { apiFetchAuthed } from '@/lib/api/client'
import type { GetAccessToken } from '@/lib/api/client'

export type ContactRecord = {
  id: string
  name: string
  phone: string
  email: string
  tags: string[]
  notes: string
  source: string
  updated_at?: string | null
}

export type ContactsOverview = {
  total: number
  contacts: ContactRecord[]
}

export async function fetchContacts(getAccessToken: GetAccessToken): Promise<ContactsOverview> {
  return apiFetchAuthed<ContactsOverview>('/v1/contacts', { method: 'GET' }, getAccessToken)
}

export async function createContact(
  getAccessToken: GetAccessToken,
  payload: { name: string; phone?: string; email?: string; notes?: string },
): Promise<{ ok: boolean; message: string }> {
  return apiFetchAuthed('/v1/contacts', { method: 'POST', body: JSON.stringify(payload) }, getAccessToken)
}

export async function deleteContact(
  getAccessToken: GetAccessToken,
  contactId: string,
): Promise<{ ok: boolean; message: string }> {
  return apiFetchAuthed(`/v1/contacts/${encodeURIComponent(contactId)}`, { method: 'DELETE' }, getAccessToken)
}

export async function importGmailContacts(
  getAccessToken: GetAccessToken,
): Promise<{ ok: boolean; message: string; total: number }> {
  return apiFetchAuthed('/v1/contacts/import/gmail', { method: 'POST' }, getAccessToken)
}

export async function importWhatsappContacts(
  getAccessToken: GetAccessToken,
): Promise<{ ok: boolean; message: string; total: number }> {
  return apiFetchAuthed('/v1/contacts/import/whatsapp', { method: 'POST' }, getAccessToken)
}
